pallets/msa/src/lib.rs

```diff
@@ -397,8 +413,55 @@ pub mod pallet {
 	#[pallet::hooks]
 	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
 		fn on_initialize(current: T::BlockNumber) -> Weight {
+			<MSAEventCount<T>>::set(0u16);
+			// allocates 1 read and 1 write for any access of `MSAEventCount` in every block
+			T::DbWeight::get().reads(1u64).saturating_add(T::DbWeight::get().writes(1u64));
 			Self::reset_virtual_bucket_if_needed(current)
 		}
+
+		fn offchain_worker(block_number: T::BlockNumber) {
+			let mut event_to_process: Vec<Event<T>> = Vec::new();
+
+			// read the event count for the block
+			let block_event_count = <MSAEventCount<T>>::get();
```

Comment:
Yep exactly my major concern about this approach, I think I am getting similar logic sorted out but agreed

Reply:
So instead of using per block counter, instead during indexing of event we can simly continue to update event count for that block in offchain index

The MSAEventCount will clear up on initialize
The MSAEvent count is a placeholder to count and index using block_number as key and a deterministic prefix

offchain_worker instead of getting the count from MSAEventCount would rather use block_number to get he count and process the events accordingly